index.js

```diff
@@ -92,13 +100,50 @@ VirtualModulesPlugin.prototype.apply = function(compiler) {
             setData(this._readFileStorage, file, [null, data.contents]);
           }.bind(this));
         }
+        if (this._virtualDirs) {
+          Object.keys(this._virtualDirs).forEach(function (dir) {
+            var data = this._virtualDirs[dir];
+            setData(this._statStorage, dir, [null, data.stats]);
+            setData(this._readdirStorage, dir, [null, data.files]);
+          }.bind(this));
+        }
       };
 
       compiler.inputFileSystem._writeVirtualFile = function(file, stats, contents) {
         this._virtualFiles = this._virtualFiles || {};
         this._virtualFiles[file] = {stats: stats, contents: contents};
         setData(this._statStorage, file, [null, stats]);
         setData(this._readFileStorage, file, [null, contents]);
+        if (this.fileSystem) {
+          var dir = path.dirname(file);
+          try {
+            this.fileSystem.statSync(dir);
+          } catch (e) {
+            var time = Date.now();
+            var dirStats = new VirtualStats({
+              dev: 8675309,
+              nlink: 0,
+              uid: 1000,
+              gid: 1000,
+              rdev: 0,
+              blksize: 4096,
+              ino: inode++,
+              mode: 16877,
+              size: stats.size,
+              blocks: Math.floor(stats.size / 4096),
+              atime: time,
+              mtime: time,
+              ctime: time,
+              birthtime: time
+            });
+            var dirData = getData(this._readdirStorage, dir);
```

Comment:
according the test case, after 'a.js' is add, then its parent virtual directory is add,too. However, when 'b.js' is triggered, readdirSync still throw a error. I don't know why, but readdirSync work properly with real directory.

Reply:
@vipcxj So its the reason to figure out what is wrong, right?

Reply:
Without combining the sub files, just override it, it seems still work. The webpack [resolver](https://github.com/webpack/enhanced-resolve/blob/master/lib/ModulesInHierachicDirectoriesPlugin.js#L36) seems not care what's in the directory, just require it exists. Just for the sake of insurance, these codes are added. 
After some study, I found `this` is `CachedInputFileSystem`. And `this.fileSystem` is `NodeJsFileSystem`. So `this.fileSystem.statSync` will always fail with the virual directory. It is the designed behaviour, in order not to interfere with the real file system, only the directories not exist should be virtualized.

Reply:
@vipcxj Then why do we use `this.fileSystem` here? We are supposed to use `compiler.inputFileSystem.statSync` then?

Reply:
it make sure `b.js` is add to the parent directory. Perhaps the webpack don't care what is under the directory now, but it may change in the future.

Reply:
@vipcxj No, you didn't understood me. I mean you are using wrong fs layer you should use CachedInputFileSystem instead.

Reply:
After trigger 'a.js', the parent directory is virsualized, So if we use `this.statSync` directly, 'b.js' will not add to the _readdirStorage.

Reply:
@vipcxj
1. Remove `if (this.fileSystem)` line and realign the code
2. Instead of `this.fileSystem.statSync` use `compiler.inputFileSystem.statSync`

Reply:
@vipcxj And we should update _readdirStorage even when stats do exist for directory, but entry to the file might not exist

Reply:
> @vipcxj And we should update _readdirStorage even when stats do exist for directory, but entry to the file might not exist

What it means? It seems that webpack not care what `_readdirStorage`  is. it care `_statsStorage` and `_readdirStorage` of directory should be exists.

Reply:
@vipcxj So I think this code should be a bit different:
```js
        var dir = path.dirname(file);
        try {
          compiler.inputFileSystem.readdirSync(dir);
        } catch (e) {
          var time = Date.now();
          var dirStats = new VirtualStats({
            dev: 8675309,
            nlink: 0,
            uid: 1000,
            gid: 1000,
            rdev: 0,
            blksize: 4096,
            ino: inode++,
            mode: 16877,
            size: stats.size,
            blocks: Math.floor(stats.size / 4096),
            atime: time,
            mtime: time,
            ctime: time,
            birthtime: time
          });
          setData(this._readdirStorage, dir, [null, []]);
          setData(this._statStorage, dir, [null, dirStats]);
        }
        var dirData = getData(this._readdirStorage, dir);
        var filename = path.basename(file);
        if (dirData[1].indexOf(filename) < 0) {
          var files = dirData[1].concat([filename]).sort();
          setData(this._readdirStorage, dir, [null, files]);
        }
```
Note:
1. We should use `readdirSync` instead of `statSync` because we are interested in readdirStorage
2. We don't need `_virtualDirs` since we create virtual dirs when we create virtual files and we update readdir storage at the same time

Reply:
@vlasenko `_virtualDirs ` is necessory. Without it statSync not work with the virtual directory. 
Look at this [code](https://github.com/webpack/enhanced-resolve/blob/master/lib/ModulesInHierachicDirectoriesPlugin.js#L36), webpack resolver use `statSync` to check the exist of parent directory. So we just make  `statSync` work property. The webpack resolver does not use `readdir` now, so it is not important now.

Reply:
@vipcxj I know, but in my code I fill stats when I detect that the directory is "virtual", so no need to keep track of virtual directories, because we already keep track of virtual files

Reply:
@vlasenko changed, It's time to go home.